executor/hash_table.go

```diff
@@ -258,101 +259,137 @@ func (c *hashRowContainer) ActionSpill() memory.ActionOnExceed {
 
 const (
 	initialEntrySliceLen = 64
-	maxEntrySliceLen     = 8 * 1024
+	maxEntrySliceLen     = 8192
 )
 
 type entry struct {
 	ptr  chunk.RowPtr
-	next entryAddr
+	next *entry
 }
 
 type entryStore struct {
 	slices [][]entry
+	cursor int
 }
 
-func (es *entryStore) init() {
-	es.slices = [][]entry{make([]entry, 0, initialEntrySliceLen)}
-	// Reserve the first empty entry, so entryAddr{} can represent nullEntryAddr.
-	reserved := es.put(entry{})
-	if reserved != nullEntryAddr {
-		panic("entryStore: first entry is not nullEntryAddr")
-	}
+func newEntryStore() *entryStore {
+	es := new(entryStore)
+	es.slices = [][]entry{make([]entry, initialEntrySliceLen)}
+	es.cursor = 0
+	return es
 }
 
-func (es *entryStore) put(e entry) entryAddr {
+func (es *entryStore) GetStore() (e *entry) {
 	sliceIdx := uint32(len(es.slices) - 1)
 	slice := es.slices[sliceIdx]
-	if len(slice) == cap(slice) {
+	if es.cursor >= cap(slice) {
 		size := cap(slice) * 2
 		if size >= maxEntrySliceLen {
 			size = maxEntrySliceLen
 		}
-		slice = make([]entry, 0, size)
+		slice = make([]entry, size)
 		es.slices = append(es.slices, slice)
 		sliceIdx++
+		es.cursor = 0
 	}
-	addr := entryAddr{sliceIdx: sliceIdx, offset: uint32(len(slice))}
-	es.slices[sliceIdx] = append(slice, e)
-	return addr
-}
-
-func (es *entryStore) get(addr entryAddr) entry {
-	return es.slices[addr.sliceIdx][addr.offset]
+	e = &es.slices[sliceIdx][es.cursor]
+	es.cursor++
+	return
 }
 
-type entryAddr struct {
-	sliceIdx uint32
-	offset   uint32
+type baseHashTable interface {
+	Put(hashKey uint64, rowPtr chunk.RowPtr)
+	Get(hashKey uint64) (rowPtrs []chunk.RowPtr)
+	Len() uint64
 }
 
-var nullEntryAddr = entryAddr{}
-
-// rowHashMap stores multiple rowPtr of rows for a given key with minimum GC overhead.
+// unsafeHashTable stores multiple rowPtr of rows for a given key with minimum GC overhead.
 // A given key can store multiple values.
 // It is not thread-safe, should only be used in one goroutine.
-type rowHashMap struct {
-	entryStore entryStore
-	hashTable  map[uint64]entryAddr
```

Comment:
Maybe this implementation has better performance.

Reply:
The performance of the `unsafeHashTable` is almost the same as the orignal `hashTable` from the benchmark.

this unsafeHashTable
```
mac:executor fzh$  GO111MODULE=on go test -bench=BenchmarkBuildHashTableForList -run=benchmark_test.go
goos: darwin
goarch: amd64
pkg: github.com/pingcap/tidb/executor
BenchmarkBuildHashTableForList/(rows:10,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-12                  18345             65191 ns/op
BenchmarkBuildHashTableForList/(rows:10,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0_1],_disk:true)-12                    9745            121827 ns/op
BenchmarkBuildHashTableForList/(rows:10,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12                   281011              4004 ns/op
BenchmarkBuildHashTableForList/(rows:10,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:true)-12                    233646              5042 ns/op
BenchmarkBuildHashTableForList/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-12                  2         654488032 ns/op
BenchmarkBuildHashTableForList/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0_1],_disk:true)-12                   1        1258633031 ns/op
BenchmarkBuildHashTableForList/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12                  112          10622517 ns/op
BenchmarkBuildHashTableForList/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:true)-12                     3         453512347 ns/op
PASS
ok      github.com/pingcap/tidb/executor        45.279s
mac:executor fzh$  GO111MODULE=on go test -bench=BenchmarkHashJoinExec -run=benchmark_test.go
goos: darwin
goarch: amd64
pkg: github.com/pingcap/tidb/executor
BenchmarkHashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-12                   1        1043568630 ns/op
BenchmarkHashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12                     7         150667029 ns/op
BenchmarkHashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:true)-12                      2         704194438 ns/op
BenchmarkHashJoinExec/(rows:5,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12                             27543             43528 ns/op
BenchmarkHashJoinExec/(rows:100000,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-12                         42          27709338 ns/op
BenchmarkHashJoinExec/(rows:100000,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12                           49          24881162 ns/op

```
master 
```
mac:executor fzh$  GO111MODULE=on go test -bench=BenchmarkBuildHashTableForList -run=benchmark_test.go
goos: darwin
goarch: amd64
pkg: github.com/pingcap/tidb/executor
BenchmarkBuildHashTableForList/(rows:10,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-12                  18372             64756 ns/op
BenchmarkBuildHashTableForList/(rows:10,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0_1],_disk:true)-12                    9505            124249 ns/op
BenchmarkBuildHashTableForList/(rows:10,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12                   296593              4083 ns/op
BenchmarkBuildHashTableForList/(rows:10,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:true)-12                    222604              4966 ns/op
BenchmarkBuildHashTableForList/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-12                  2         613439571 ns/op
BenchmarkBuildHashTableForList/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0_1],_disk:true)-12                   2         928064504 ns/op
BenchmarkBuildHashTableForList/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12                  117           9789629 ns/op
BenchmarkBuildHashTableForList/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:true)-12                     3         438074455 ns/op
PASS
ok      github.com/pingcap/tidb/executor        47.873s

mac:executor fzh$   GO111MODULE=on go test -bench=BenchmarkHashJoinExec -run=benchmark_test.go
goos: darwin
goarch: amd64
pkg: github.com/pingcap/tidb/executor
BenchmarkHashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-12                   1        1121902387 ns/op
BenchmarkHashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12                     7         150140849 ns/op
BenchmarkHashJoinExec/(rows:100000,_cols:[bigint(20)_var_string(5)],_concurency:4,_joinKeyIdx:_[0],_disk:true)-12                      2         659220502 ns/op
BenchmarkHashJoinExec/(rows:5,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12                             26731             44061 ns/op
BenchmarkHashJoinExec/(rows:100000,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0_1],_disk:false)-12                         45          26341352 ns/op
BenchmarkHashJoinExec/(rows:100000,_cols:[bigint(20)_double],_concurency:4,_joinKeyIdx:_[0],_disk:false)-12                           50          23952114 ns/op
PASS
ok      github.com/pingcap/tidb/executor        23.740s
```